### PR TITLE
Allow convention-oriented topic mappings to be set via environment variables

### PR DESCRIPTION
**Pull request by jessemyers**
We increasingly implement configuration via environment variables, but we our
pubsub media type conventions tend to use naming schemas that can't be expressed
using environment variables. In particular, using previous logic, we'd want to
generate a configuration dictionar with something like:

    {
      "sns_topic_arns": {
        "mappings": {
          "application/vnd.globality.pubsub._.created.foo_bar.baz": "topic-name",
        }
      }
    }

While the microcosm environment variable loader can handle nested dictionaries
(via double underscores), we cannot express '/' or '.' because they aren't legal
characters.

We solve this problem in two ways:

 1. We add support for the convention-driven media types using alternative keys (avoiding '/')
 2. We (re)collapse double underscores within these convention keys (avoiding '.')

That is, we implement the above mapping using:

    {
      "sns_topic_arns": {
        "created": {
          "foo_bar": {
            "baz": "topic-name",
          },
        }
      }
    }

### CODE REVIEW

**Comment by KensoDev** [comment]
LGTM

File: microcosm_pubsub/producer.py [hunk 1]
```diff
@@ -109,6 +111,33 @@ def __exit__(self, type, value, traceback):
...
+        else:
+            yield separator.join([prefix, key]), value
+
+
+def iter_topic_mappings(dct):
+    for key, value in dct.items():
+        if isinstance(value, string_types):
+            yield key, value
+        else:
+            for key, value in collapse_dict(value, key):
```
Comment (pelegdvir) [inline microcosm_pubsub/producer.py]: Rename (key, value) names (in the left side) - you already used the key, value names

Comment (jessemyers) [inline microcosm_pubsub/producer.py]: You can't return from inside a generator; you have to yield.

I fixed the naming.

File: microcosm_pubsub/tests/test_producer.py [hunk 2]
```diff
@@ -115,6 +118,42 @@ def loader(metadata):
...
+                ),
+                baz=dict(
+                    foo=dict(
+                        bar="foo",
+                    ),
+                ),
+            )
+        )
+    )
+    assert_that(result, is_(equal_to({
```
Comment (pelegdvir) [inline microcosm_pubsub/tests/test_producer.py]: Can you expand this test - add  to the end something like `"bar.foo2": "baz2"`

**Pull request closed without merging**